mopro-ffi/src/ashlang/mod.rs

```diff
@@ -0,0 +1,70 @@
+use std::fs;
+
+use ashlang::AshlangProver;
+use ashlang::SpartanProver;
+
+use super::GenerateProofResult;
+
+#[macro_export]
+macro_rules! ashlang_spartan_app {
+    () => {
+        fn generate_ashlang_spartan_proof(
+            ar1cs_path: String, // path to ar1cs file
+            secret_inputs: Vec<String>,
+        ) -> Result<mopro_ffi::GenerateProofResult, mopro_ffi::MoproError> {
+            mopro_ffi::ashlang::prove(&ar1cs_path, secret_inputs).map_err(|e| {
+                mopro_ffi::MoproError::AshlangError(
+                    "error generating ashlang spartan proof".to_string(),
+                )
+            })
+        }
+
+        fn verify_ashlang_spartan_proof(
+            ar1cs_path: String,
+            proof: Vec<u8>,
+        ) -> Result<bool, mopro_ffi::MoproError> {
+            mopro_ffi::ashlang::verify(&ar1cs_path, proof).map_err(|e| {
+                mopro_ffi::MoproError::AshlangError("error verifying proof".to_string())
+            })
+        }
+    };
+}
+
+/// Generates a spartan proof from an ar1cs file compiled
+/// with the ashlang compiler.
+pub fn prove(
+    ar1cs_path: &str, // path to ar1cs file
+    secret_inputs: Vec<String>,
+) -> anyhow::Result<GenerateProofResult> {
+    let ir_source = fs::read_to_string(ar1cs_path)?;
+    // we pass an empty vec for public inputs because
+    // they are not supported in the ashlang spartan prover
+    // outputs are public and should be used instead
+    let proof = SpartanProver::prove_ir(&ir_source, vec![], secret_inputs.clone())?;
+
+    Ok(GenerateProofResult {
+        proof: bincode::serialize(&proof)?,
+        inputs: bincode::serialize(&secret_inputs)?,
+    })
+}
+
+/// Verifies a spartan proof from an ar1cs file
+pub fn verify(ar1cs_path: &str, proof: Vec<u8>) -> anyhow::Result<bool> {
+    let ir_source = fs::read_to_string(ar1cs_path)?;
+    let p = bincode::deserialize(&proof)?;
+    ashlang::SpartanProver::verify(&ir_source, p)
+    // Ok(true)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_ashlang_prove_verify() -> anyhow::Result<()> {
+        let ar1cs_path = "../test-vectors/ashlang/example.ar1cs".to_string();
```

Comment:
I am thinking about to use a more meaningful circuit
e.g. `multiplier2` so that we can compare with circom prover

Reply:
What do you want to compare?

Reply:
like why people choose from circom to spartan 
In this example, the multiplier takes 0.02s to prove but the `ashlang_spartan_prover` takes 1s to prove
it could be hard to convince people what is the advantage of using `ashlang_spartan_prover`

Reply:
Ah, the purpose of this implementation is not to convince someone to use spartan over groth16. IMO benchmarks should happen independently/separately from this integration.

I don't care much about the spartan prover though, it's ECC based so given the choice I'd use the stark prover (though it doesn't support 32 bit architectures right now).

Reply:
I am thinking about the `example` circuit is neither meaningful (people won't know it proves for a hash or some operations) nor fast
like `multiplier2` and `fibonacci` is the most simple/fast circuit in circom and halo2
can we switch to a fast and meaningful circuit (e.g. `multiplier2`)?

Reply:
looking forward to a stark prover 😄

Reply:
Public inputs and outputs are not implemented in the [ashlang spartan prover](https://github.com/chancehudson/ashlang/blob/main/ashlang/src/provers/spartan.rs) yet (PR's are certainly welcome though). I'm a bit short on bandwidth right now and need to switch back to the LaBRADOR implementation I'm working on. I'd recommend merging this and keeping it behind a feature flag/out of the docs until the public inputs/outputs are supported.

Reply:
Essentially all that needs to happen is the r1cs format is transformed to match what the spartan prover expects. Public inputs and outputs are already supported by the language/compiler they just need to be bridged to the spartan prover.